File: app/src/main/java/com/kaciula/archiman/data/local/prefs/util/GsonPreference.java
package com.kaciula.archiman.data.local.prefs.util;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.kaciula.archiman.util.Strings;

public class GsonPreference<T> {
    private final SharedPreferences preferences;
    private final String key;
    private final T defaultValue;
    private final Gson gson;
    private Class<T> type;

    public GsonPreference(SharedPreferences preferences, String key, Class<T> type, Gson gson) {
        this(preferences, key, null, type, gson);
    }

    public GsonPreference(SharedPreferences preferences, String key, T defaultValue, Class<T>
            type, Gson gson) {
        this.preferences = preferences;
        this.key = key;
        this.defaultValue = defaultValue;
        this.type = type;
        this.gson = gson;
    }

    public T get() {
        String value = preferences.getString(key, "");
        if (Strings.isBlank(value)) {
            return defaultValue;
        }
        return gson.fromJson(value, type);
    }

    public boolean isSet() {
        return preferences.contains(key);
    }

    public void set(T value) {
        if (value != null) {
            preferences.edit().putString(key, gson.toJson(value, type)).apply();
        } else {
            preferences.edit().putString(key, null).apply();
        }
    }

    public void delete() {
        preferences.edit().remove(key).apply();
    }
}

File: app/src/main/java/com/kaciula/archiman/data/local/prefs/util/GsonPreference.kt
package com.kaciula.archiman.data.local.prefs.util

import android.content.SharedPreferences
import com.google.gson.Gson

class GsonPreference<T>(private val preferences: SharedPreferences, private val key: String,
                        private val defaultValue: T, private val type: Class<T>,
                        private val gson: Gson) {

    fun get(): T {
        val value = preferences.getString(key, "")
        if (value.isBlank()) {
            return defaultValue
        }
        return gson.fromJson(value, type)
    }

    fun isSet(): Boolean {
        return preferences.contains(key)
    }

    fun set(value: T) {
        if (value != null) {
            preferences.edit().putString(key, gson.toJson(value, type)).apply()
        } else {
            preferences.edit().putString(key, null).apply()
        }
    }

    fun delete() {
        preferences.edit().remove(key).apply()
    }
}
